### PR TITLE
Adkrawcz/imgui

### PR DESCRIPTION
**Pull request by adamkrawczyk**
## Updates
1. Integration of ImGui with ROS2 Sensors: Implemented ImGui to enhance the user interface for ROS2 sensors.
1. Custom Behavior Link via Bus: Added a bus system to link custom behaviors to button clicks.
1. GNSS Post-Processing Bus: Established a bus for GNSS post-processing.
1. Standard ImGui Element Creation: Developed a standard ImGui element for consistent user interface elements.

## Detailed Description
ROS 2 Sensor Publishing Control: Introduced an option to enable or disable data publishing from ROS 2 sensors. While the component remains active and the topic is visible, this feature allows the emulation of a sensor fault by stopping data publication on the topic. This is useful for simulating scenarios where the sensor is faulty and does not publish any data.
To have the sensor visible in the ImGui window the user needs to add a component SensorImGuiComponent to the entity where the sensors are added.

## User Experience

1. Add SensorImGuiComponent to the entity with the sensors you want to control.
2. When the game is launched press the `~` to start ImGui 
3. The user will see ` ROS 2 ImGui` tab on ImGui main menu 

Screenshot from 2024-01-02 12-49-42

4. After clicking the list of the link names to which the `SensorImGuiComponent(s)` were added will pop up. 

Screenshot from 2024-01-02 12-49-59

5. The user can click the link names to add sensors from this link to debug window. When the link is selected to be visible the tick mark is on the right of the name.

Screenshot from 2024-01-02 12-55-49

6. The basic SensorImGuiComponent allows to disable/enable publishing and this is as follows: 

Screenshot from 2024-01-02 12-50-28

7. When the button is clicked the text will change to Enable and no data is published on the topic(s) shown above the button. If the sensor publishes on more than one topic (e.x. camera_image and camera_info) all topic names will be shown and a button click disables/enables publishing on all of them. 

Screenshot from 2024-01-02 12-50-34

### CODE REVIEW

**Comment by adamkrawczyk** [comment]
To work correctly it needs interface from https://github.com/o3de/o3de-extras/pull/639

File: repo.json [hunk 1]
```diff
@@ -34,6 +34,34 @@
...
+            "requirements": "Requires ROS 2 installation (supported distributions: Humble). Source your workspace before building the Gem",
+            "documentation_url": "TODO",
+            "dependencies": [],
+            "repo_uri": "",
+            "compatible_engines": [],
+            "engine_api_dependencies": [],
+            "restricted": "RobotecImGui"
         }
     ]
 }
```
Comment (michalpelka) [inline repo.json:67]: nit : newline

File: repo.json [hunk 2]
```diff
@@ -34,6 +34,34 @@
...
+            "gem_name": "RobotecImGui",
+            "version": "1.0.0",
+            "platforms": [
+                "Linux"
+            ],
+            "license": "Apache-2.0",
+            "display_name": "Robotec ImGui",
+            "license_url": "https://opensource.org/licenses/Apache-2.0",
+            "origin": "RobotecAI",
+            "origin_url": "TODO",
```
Comment (michalpelka) [inline repo.json:48]: Give url to git repo here.

File: Gems/RobotecImGui/gem.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
+    "gem_name": "RobotecImGui",
+    "version": "1.0.0",
+    "display_name": "RobotecImGui",
+    "license": "License used i.e. Apache-2.0 or MIT",
```
Comment (michalpelka) [inline Gems/RobotecImGui/gem.json:5]: Fill license (e.g. prorietary)

File: Gems/RobotecImGui/gem.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    "license_url": "Link to the license web site i.e. https://opensource.org/licenses/Apache-2.0",
+    "origin": "The name of the originator or creator",
```
Comment (michalpelka) [inline Gems/RobotecImGui/gem.json:7]: url to repo.

File: Gems/RobotecImGui/gem.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    "origin_url": "The website for this Gem",
+    "type": "Code",
+    "summary": "A short description of this Gem",
```
Comment (michalpelka) [inline Gems/RobotecImGui/gem.json:10]: write description.

File: Gems/RobotecImGui/gem.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        "Gem"
+    ],
+    "user_tags": [
+        "RobotecImGui"
+    ],
+    "platforms": [
+        ""
+    ],
+    "icon_path": "preview.png",
+    "requirements": "Notice of any requirements for this Gem i.e. This requires X other gem",
```
Comment (michalpelka) [inline Gems/RobotecImGui/gem.json:21]: Fill requirments.

File: Gems/RobotecImGui/Code/Source/ImGui/SensorImGui.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                ? sharedSensorComponentTickBased->GetSensorConfiguration()
+                : sharedSensorComponentPhysicsBased->GetSensorConfiguration();
+
+            AZStd::string label;
+            for (const auto& [key, value] : sensorConfiguration.m_publishersConfigurations)
+            {
+                label += "Topic: " + value.m_topic + "\n";
+            }
+
+            AZStd::string buttonText = sensorConfiguration.m_publishingEnabled ? "Disable Publishing" : "Enable Publishing";
```
Comment (michalpelka) [inline Gems/RobotecImGui/Code/Source/ImGui/SensorImGui.cpp:65]: add const.

File: Gems/RobotecImGui/Code/Source/ImGui/SensorImGui.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            return {};
+        }
+
+        for (const auto& component : childComponents)
+        {
+            std::shared_ptr<ROS2::ROS2SensorComponentBase<ROS2::TickBasedSource>> sharedSensorComponentTickBased;
+            std::shared_ptr<ROS2::ROS2SensorComponentBase<ROS2::PhysicsBasedSource>> sharedSensorComponentPhysicsBased;
+
+            if (auto* tickBased = azrtti_cast<ROS2::ROS2SensorComponentBase<ROS2::TickBasedSource>*>(component))
```
Comment (michalpelka) [inline Gems/RobotecImGui/Code/Source/ImGui/SensorImGui.cpp:31]: Please move this code to another method for both implementations. At this level leave common logic.

File: Gems/RobotecImGui/Code/Source/Clients/RobotecImGuiSystemComponent.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        static void GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided);
+        static void GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& incompatible);
+        static void GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required);
+        static void GetDependentServices(AZ::ComponentDescriptor::DependencyArrayType& dependent);
+
+        RobotecImGuiSystemComponent();
+        ~RobotecImGuiSystemComponent();
+
+    protected:
+        ////////////////////////////////////////////////////////////////////////
```
Comment (michalpelka) [inline Gems/RobotecImGui/Code/Source/Clients/RobotecImGuiSystemComponent.h:29]: Please adjust base class names to follow convention.

**Pull request closed without merging**